### PR TITLE
feat(security): add http basic auth (fix #6)

### LINKED ISSUE #6: Authentication
I think it would be much more convenient if we can add authentication like HTTP Authentication, query parameter,IP restriction etc.

### PR DESCRIPTION
**Pull request by eternal-flame-AD**
I added an internal package named auth which exports a map[string]Method containing available authentication methods. Each authentication should implement the auth.Method interface so that they can be used as an authentication method.

### CODE REVIEW

File: pkg/auth/basic.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+type basicAuth struct {
+	username   string
+	password   string
+	debug      bool
+	authheader string
+}
+
+func (c *basicAuth) Init(debug bool) {
+	c.debug = debug
+	logger.Warning.Println("\u001B[33mBasic Auth: Debug mode enabled. Might Leak sentitive information in log output.\u001B[0m")
```
Comment (eternal-flame-AD) [inline pkg/auth/basic.go]: My fault:(

Gotta fix this later.

File: config.go [hunk 2]
```diff
@@ -1,31 +1,49 @@
...
-	Debug:      flag.Bool("debug", getBoolEnv("DEBUG", false), "Output debug logs"),
-	Timeout:    flag.Int("timeout", getIntEnv("HOOK_TIMEOUT", 10), "Hook maximum delay before timeout (in second)"),
-	ScriptDir:  flag.String("scripts", getEnv("SCRIPTS_DIR", "scripts"), "Scripts directory"),
+	ListenAddr:     flag.String("listen", getEnv("LISTEN_ADDR", ":8080"), "HTTP service address (e.g.address, ':8080')"),
+	NbWorkers:      flag.Int("nb-workers", getIntEnv("NB_WORKERS", 2), "The number of workers to start"),
+	Debug:          flag.Bool("debug", getBoolEnv("DEBUG", false), "Output debug logs"),
+	Timeout:        flag.Int("timeout", getIntEnv("HOOK_TIMEOUT", 10), "Hook maximum delay before timeout (in second)"),
+	ScriptDir:      flag.String("scripts", getEnv("SCRIPTS_DIR", "scripts"), "Scripts directory"),
+	Authentication: flag.String("auth", getEnv("AUTH", "none"), ""),
+	AuthenticationParam: flag.String("authparam", getEnv("AUTH_PARAM", ""), func() string {
```
Comment (ncarlier) [inline config.go]: Can you use this parameter name: `auth-param` ?

File: main.go [hunk 3]
```diff
@@ -34,13 +35,29 @@ func main() {
...
+	}
+
 	level := "info"
 	if *config.Debug {
 		level = "debug"
 	}
 	logger.Init(level)
 
 	logger.Debug.Println("Starting webhookd server...")
+	logger.Info.Println("Using Authentication:", name)
```
Comment (ncarlier) [inline main.go]: Can you use debug level for the logger?

File: pkg/auth/basic.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+}
+
+// BasicAuth HTTP Basic Auth implementation
+func (c *basicAuth) Middleware() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if username, password, ok := r.BasicAuth(); ok && username == c.username && password == c.password {
+				if c.debug {
+					logger.Debug.Printf("HTTP Basic Auth: %s:%s PASSED\n", username, password)
```
Comment (ncarlier) [inline pkg/auth/basic.go]: Can you remove this line. I think we should not log credentials.

File: pkg/auth/basic.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+			} else if !ok {
+				if c.debug {
+					logger.Debug.Println("HTTP Basic Auth: Auth header not present.")
+				}
+				w.Header().Add("WWW-Authenticate", c.authheader)
+				w.WriteHeader(401)
+				w.Write([]byte("Authentication required."))
+			} else {
+				if c.debug {
+					logger.Debug.Printf("HTTP Basic Auth: Invalid credentials: %s:%s \n", username, password)
```
Comment (ncarlier) [inline pkg/auth/basic.go]: Maybe just log the username: "Invalid credentials for user xxx."

**Comment by ncarlier** [comment]
Thanks a lot!

**Pull request merged**